pkg/controller/master/migration/vmim_controller.go

```diff
@@ -88,3 +92,176 @@ func (h *Handler) syncVM(vmi *kubevirtv1.VirtualMachineInstance) error {
+func (h *Handler) scaleResourceQuotaWithVMI(vmi *kubevirtv1.VirtualMachineInstance) error {
+	// Scale ResourceQuota through VMI resource specifications
+	rqs, err := h.rqCache.List(vmi.Namespace, labels.Everything())
+	if err != nil || len(rqs) == 0 {
+		return err
+	}
+
+	rqCpy := rqs[0].DeepCopy()
+	if _, ok := h.isVMIExistsMigrationVMsAnnotation(rqCpy, vmi.Name); ok {
+		return nil
+	}
+
+	needUpdate, toUpdate, rl := rqutils.CalculateScaleResourceQuotaWithVMI(rqCpy, vmi)
+	if !needUpdate {
+		return nil
+	}
+
+	// Update resource quota
+	if err := h.setMigrationVMsAnnotation(toUpdate, vmi.Name, rl); err != nil {
+		return err
+	}
+	_, err = h.rqs.Update(toUpdate)
+	return err
+}
+
+// restoreResourceQuota restores the resource quota when the migration is completed
+func (h *Handler) restoreResourceQuota(vmim *kubevirtv1.VirtualMachineInstanceMigration, vmi *kubevirtv1.VirtualMachineInstance) error {
+	if vmim.Status.Phase != kubevirtv1.MigrationFailed && vmim.Status.Phase != kubevirtv1.MigrationSucceeded {
+		return nil
+	}
+
+	// If the namespace is not managed by the resource quota, skip scaling
+	if exist, err := h.isNamespaceManagedByResourceQuota(vmi.Namespace); exist == false && err == nil {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to check if the namespace is managed by the resource quota: %v", err)
+	}
+
+	// Restore ResourceQuota through VMI resource specifications
+	if err := h.restoreResourceQuotaWithVMI(vmi); err != nil {
+		return fmt.Errorf("failed to restore resource quota with vmi: %v", err)
+	}
+
+	return nil
+}
+
+// restoreResourceQuotaWithVMI restores the resource quota when the migration is completed
+func (h *Handler) restoreResourceQuotaWithVMI(vmi *kubevirtv1.VirtualMachineInstance) error {
+	// Restore ResourceQuota through VMI resource specifications
+	rqs, err := h.rqCache.List(vmi.Namespace, labels.Everything())
+	if err != nil || len(rqs) == 0 {
+		return err
+	}
+
+	rqCpy := rqs[0].DeepCopy()
+	rl, ok := h.isVMIExistsMigrationVMsAnnotation(rqCpy, vmi.Name)
+	if !ok {
+		return nil
+	}
+
+	needUpdate, toUpdate := rqutils.CalculateRestoreResourceQuotaWithVMI(rqCpy, vmi, rl)
+	if !needUpdate {
+		return nil
+	}
+
+	// Update resource quota
+	if err := h.removeMigrationVMsAnnotation(toUpdate, vmi.Name); err != nil {
+		return err
+	}
+	_, err = h.rqs.Update(toUpdate)
+	return err
+}
+
+func (h *Handler) getMigrationVMsAnnotation(rq *corev1.ResourceQuota) (map[string]corev1.ResourceList, error) {
+	migratingVMs := make(map[string]corev1.ResourceList)
+	if rq.Annotations != nil {
+		if v, ok := rq.Annotations[util.AnnotationMigratingVMs]; ok && v != "" {
+			if err := json.Unmarshal([]byte(rq.Annotations[util.AnnotationMigratingVMs]), &migratingVMs); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return migratingVMs, nil
+}
+
+func (h *Handler) updateMigrationVMsAnnotation(rq *corev1.ResourceQuota, migratingVMs map[string]corev1.ResourceList) error {
```

Comment:
we are doing a lot of work to marshal/unmarshal map[vmName]ResourceList into json and then updating the annotation. Is there a reason we cant just update the ResourceQuote annotations directly with something like `harvesterhci.io/vmName: jsonOfResourceList` it would make it a lot more efficient to find if there is already a quota for VM being tracked. The current logic works just feel we could make it simpler. @bk201 any thoughts about the same?